### ISSUE TITLE
ConvNets performance tuning

### ISSUE THREAD

**Issue by gdamaskinos** [mortendahl/tf-encrypted#339]
I am considering using this library for a research project and would like to get the best performance (inference latency) possible before I (hopefully) manage to run a large model (e.g., resnet-152) on imagenet.

I implemented a working+simplified version of [bench_conv2d_sigmoid.py](https://github.com/mortendahl/tf-encrypted/blob/master/examples/bench_conv2d_sigmoid.py), including the same computations on vanilla tensorflow.
<details><summary>Code</summary>
<p>

```python
import sys
import time
from functools import reduce
import tensorflow as tf
import tf_encrypted as tfe

config = tfe.LocalConfig([
    'server0',
    'server1',
    'crypto-producer',
    'weights-provider',
    'prediction-client'
])

tfe.set_config(config)
#tfe.set_protocol(tfe.protocol.Pond())
tfe.set_protocol(tfe.protocol.SecureNN())

input_shape1 = [1, 3, 32, 32]
input_shape2 = [1, 3, 192, 192]
conv1_fshape = [3, 3, 3, 32]

def provide_input_conv1weights() -> tf.Tensor:
    w = tf.random_normal(shape=conv1_fshape, dtype=tf.float32)
    return tf.Print(w, [w], message="w1:")

def provide_input_prediction() -> tf.Tensor:
    x = tf.random_normal(shape=input_shape1, dtype=tf.float32)
    return tf.Print(x, [x], message="x:")

def provide_input_prediction2() -> tf.Tensor:
    x = tf.random_normal(shape=input_shape2, dtype=tf.float32)
    return tf.Print(x, [x], message="x:")

def receive_output(tensor: tf.Tensor) -> tf.Operation:
    return tf.Print(tensor, [tensor, tf.shape(tensor)], message="output:")

def tfe_inference(x):
  conv1 = tfe.layers.Conv2D(x.shape.as_list(), conv1_fshape, 1, "SAME")
  initial_w_conv1 = tfe.define_private_input('weights-provider',
                                             provide_input_conv1weights)
  conv1.initialize(initial_w_conv1)
  x = conv1.forward(x)
  relu1 = tfe.layers.activation.Relu(x.shape.as_list())
  x = relu1.forward(x)
  pool1 = tfe.layers.pooling.MaxPooling2D(x.shape.as_list(), pool_size=2,
                                          strides=2, padding='SAME')
  x = pool1.forward(x)
  shape = reduce(lambda x,y: x*y, x.shape.as_list())
  x = x.reshape([shape, -1])

  dense = tfe.layers.Dense(x.shape, 2)
  dense.initialize()
  x = dense.forward(x)

  return x

def tf_inference(x):
  x = tf.nn.conv2d(x, provide_input_conv1weights(), strides=[1,1,1,1], padding="SAME")
  x = tf.nn.relu(x)
  x = tf.nn.max_pool(x, ksize=[1,2,2,1], strides=[1,2,2,1], padding='SAME')

  shape1 = reduce(lambda x,y: x*y, x.shape.as_list())
  x = tf.reshape(x, shape=[shape1, -1])
  return tf.layers.dense(x, 2)


for input_func, m in zip([provide_input_prediction,
                          provide_input_prediction2],
                         ["small input", "large input"]):
  print(m)
  with tf.Graph().as_default():
    print("TFE")
    x = tfe.define_private_input('prediction-client', input_func)
    y = tfe_inference(x)
    prediction_op = tfe.define_output('prediction-client', [y], receive_output)

    with tfe.Session(config=config) as sess:
        print("Initialize tensors")
        sess.run(tf.global_variables_initializer(), tag='init')

        print("Predict")
        for i in range(3):
          t = time.time()
          sess.run(prediction_op, tag='prediction')
          print("Inference time: %g" % (time.time() - t))


  with tf.Graph().as_default():
    print("TF")
    x = input_func()
    x = tf.transpose(x, (0,2,3,1)) # NHWC

    prediction_op = tf_inference(x)

    with tf.Session() as sess:
        print("Initialize tensors")
        sess.run(tf.global_variables_initializer())

        print("Predict")
        for i in range(3):
          t = time.time()
          sess.run(prediction_op)
          print("Inference time: %g" % (time.time() - t))
```

</p>
</details>

I did some tests by using different options and here are the performance results I get:

| Input size | Implementation | Prediction#1 | Prediction#2 | Prediction#3 |
| --- | --- | --- | --- | --- |
| 3x32x32 | tf-encrypted | 44.4 s | 2.8 s | 2.8 s |
| 3x32x32 | tf | 0.01 s | 0.001 s | 0.001 s |
| 3x192x192 | tf-encrypted | 1146.9 s | 138.1 s | 136.7 s |
| 3x192x192 | tf | 0.03 s | 0.02 s | 0.02 s |

There are 2 problems:
1. the inference latency of tf-encrypted (for the second and third prediction) is 3-5 orders of magnitude bigger than vanilla tensorflow. 
2. tf-encrypted is scaling very badly with increasing input size.

Of course there should be an overhead comparing to vanilla TF, but this difference is too big to even hope for [a reasonable performance for a large network](https://dawn.cs.stanford.edu/benchmark/#imagenet-inference-time).
I would really appreciate any ideas/suggestions for tuning-efficiently using your library?

**Comment by mortendahl** [comment]
Hi @gdamaskinos,

Amazing to hear you're considering tf-encrypted! Let's see if we can't get those numbers down: computing on encrypted data is inherently more expensive (say currently at least two orders of magnitude) but I believe what you're seeing can be improved.

I'll continue to investigate your model and get back to you next week but in the meantime here are a few things to try:

- First thing is to make sure you're using the latest version of tf-encrypted as we had a bug very recently around sampling of randomness that made everything significantly slower. If you simply pull and use the `master` branch you should be good (run `pip install -e .` to install it).

- Then make sure you're using a version of TensorFlow that supports int64 operations, which can be verified via `tfe.config.tensorflow_supports_int64()`. You can get a custom version [here](https://github.com/mortendahl/tf-encrypted/#securenn-int64-support) until the TensorFlow patch becomes part of the official release.

With these a quick test shows ~50s for the large models on my laptop, but note that using `LocalConfig` might not give an accurate idea of the performance you get from running in an actual cluster; let me get back to you on how to test this in a cluster setting.

There has been some prior experiments with larger networks (a variant of VGG16) that @yanndupis might be able to provide a quick outline of, including any optimizations used.

**Comment by yanndupis** [comment]
Hello @gdamaskinos,

Thank you very much for your question. As a reference point, we benchmarked a variant of VGG16 model. It was the same architecture except we got rid of the largest fully connected layer (~100 million parameters) and we predicted 2 classes. The model contained about 21 million parameters. 

With 96 cpus on each instance (crypto producer, server 0, server 1 etc) it was taking about 1min and 4 seconds. 

If you are using large models with ReLU and MaxPooling layer, it's important to use the SecureNN protocol. The other protocol, Pond, doesn't support comparison so you can't use  MaxPooling with it and ReLU is approximated. With the SecureNN protocol you will have the MaxPooling layer with the exact ReLU. 

Do not hesitate to ask if you need further precisions. Thank you.

**Comment by gdamaskinos** [comment]
Thank you very much for your rapid responses and help.
I modified the code a bit by following your advice (remote setup; employing a frozen graph instead of defining the graph in tfe).

<details><summary>Code v2</summary>
<p>

```python
import sys
import time
import numpy as np

from functools import partial
from functools import reduce
import tensorflow as tf
import tf_encrypted as tfe
from tensorflow.python.framework import graph_util, graph_io

print(tfe.__file__)
print("Support for int64: ", tfe.config.tensorflow_supports_int64())

tf.set_random_seed(42)
np.random.seed(42)

small = False
if small:
  print("Running CNN with small input")
else:
  print("Running CNN with large input")

config = tfe.LocalConfig([
    'server0',
    'server1',
    'crypto-producer',
    'weights-provider',
    'prediction-client'
])

#config = tfe.RemoteConfig([
#  ('server0', 'server0:4441'),
#  ('server1', 'server1:4442'),
#  ('crypto-producer', 'server2: 4443'),
#  ('weights-provider', 'server3: 4444'),
#  ('prediction-client', 'server2:4442')
#  ], log_device_placement=False)

if len(sys.argv) > 1:
  if isinstance(config, tfe.LocalConfig):
      raise Exception("You can launch a configured server only with a remote configuration")
  player_name = str(sys.argv[1])
  server = config.server(player_name)
  server.start()
  server.join()

tfe.set_config(config)
tfe.set_protocol(tfe.protocol.SecureNN())

input_shape1 = [1, 3, 32, 32]
input_shape2 = [1, 3, 192, 192]
conv1_fshape = [3, 3, 3, 32]

def provide_input_conv1weights() -> tf.Tensor:
  w = tf.constant(np.random.normal(size=conv1_fshape), dtype=tf.float32)
  return w

def provide_input_denseweights(shape) -> tf.Tensor:
  w = tf.constant(np.random.normal(size=[shape, 2]), dtype=tf.float32)
  return w

def provide_input_prediction() -> tf.Tensor:
  x = tf.random_normal(shape=input_shape1, dtype=tf.float32, seed=42)
  return x

def provide_input_prediction2() -> tf.Tensor:
  x = tf.random_normal(shape=input_shape2, dtype=tf.float32)
  return x

def receive_output(tensor: tf.Tensor) -> tf.Operation:
  return tf.Print(tensor, [tensor, tf.shape(tensor)], message="output:")

def tfe_inference(x):
  conv1 = tfe.layers.Conv2D(x.shape.as_list(), conv1_fshape, 1, "SAME")
  initial_w_conv1 = tfe.define_private_input('weights-provider',
                                             provide_input_conv1weights)
  conv1.initialize(initial_w_conv1)
  x = conv1.forward(x)
  relu1 = tfe.layers.activation.Relu(x.shape.as_list())
  x = relu1.forward(x)
  pool1 = tfe.layers.pooling.MaxPooling2D(x.shape.as_list(), pool_size=2,
                                          strides=2, padding='SAME')

  x = pool1.forward(x)
  shape = reduce(lambda x,y: x*y, x.shape.as_list())
  x = x.reshape([-1, shape])

  part = partial(provide_input_denseweights, shape)
  initial_w_dense = tfe.define_private_input('weights-provider', part)
  x = tfe.matmul(x, initial_w_dense)
  return x

def tf_inference(x, data_format="NHWC"):
  """Tensorflow only supports inference with NHWC; tfe convert needs NCHW"""
  x = tf.nn.conv2d(x, provide_input_conv1weights(), strides=[1,1,1,1],
                   padding="SAME", data_format=data_format)
  x = tf.nn.relu(x)
  x = tf.nn.max_pool(x, ksize=[1,2,2,1], strides=[1,2,2,1], padding='SAME',
                     data_format=data_format)

  shape = reduce(lambda x,y: x*y, x.shape.as_list())
  x = tf.reshape(x, shape=[-1, shape])
  return tf.matmul(x, provide_input_denseweights(shape))

def export_cnn(x):
  x = tf_inference(x, data_format='NCHW')

  pred_node_names = ["output"]
  tf.identity(x, name=pred_node_names[0])

  with tf.Session() as sess:
    constant_graph = graph_util.convert_variables_to_constants(sess,
                                                               sess.graph.as_graph_def(),
                                                               pred_node_names)
  frozen = graph_util.remove_training_nodes(constant_graph)
  graph_io.write_graph(frozen, ".", "bench_cnn.pb", as_text=False)

if small:
  input_func = provide_input_prediction
else:
  input_func = provide_input_prediction2

with tf.Graph().as_default():
  print("Exporting graph...")
  x = input_func()
  x = tf.placeholder(tf.float32, shape=x.shape)
  export_cnn(x)

with tf.Graph().as_default():
  print("Loading graph...")
  with tf.gfile.GFile("bench_cnn.pb", 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Graph().as_default():
  print("TFE")
  c = tfe.convert.Converter(config, tfe.get_protocol(),
                            config.get_player('weights-provider'))
  y = c.convert(graph_def, tfe.convert.register(),
                config.get_player('prediction-client'), input_func)

  #x = tfe.define_private_input('prediction-client', input_func)
  #y = tfe_inference(x)

  prediction_op = tfe.define_output('prediction-client', [y], receive_output)

  with tfe.Session(config=config) as sess:
      print("Initialize tensors")
      sess.run(tf.global_variables_initializer(), tag='init')

      print("Predict")
      for i in range(3):
        t = time.time()
        sess.run(prediction_op, tag='prediction')
        print("Inference time: %g" % (time.time() - t))


with tf.Graph().as_default():
  print("TF")
  x = input_func()
  x = tf.transpose(x, (0,2,3,1)) # NHWC

  prediction_op = tf_inference(x)

  with tf.Session() as sess:
      print("Initialize tensors")
      sess.run(tf.global_variables_initializer())

      print("Predict")
      for i in range(3):
        t = time.time()
        sess.run(prediction_op)
        print("Inference time: %g" % (time.time() - t))
```

</p>
</details>


I run the benchmark on a better server (after building the secure random shared objects) and got significantly better results than the previous ones I posted for both tf and tfe.
The code was tested with release 0.4.0.

| Input size | Implementation | Prediction#1 | Prediction#2 | Prediction#3 |
| --- | --- | --- | --- | --- |
| 3x32x32 | tfe-local | 19.6 s | 0.3 s | 0.3 s |
| 3x32x32 | tfe-local (int64 ops) | 0.73 s | 0.2 s | 0.2 s |
| 3x32x32 | tfe-remote | 21.8 s | 1.3 s | 1.2 s |
| 3x32x32 | tfe-remote (int64 ops) | 2 s | 1.1 s | 1 s |
| 3x32x32 | tf | 0.006 s | 0.0006 s | 0.0006 s |
| 3x192x192 | tfe-local | 140.3 s | 13.9 s | 13.5 s |
| 3x192x192 | tfe-local (int64 ops) | 10.5 s | 8.8 s | 8.4 s |
| 3x192x192 | tfe-remote | 168.7 s | 39.9 s | 40.7 s |
| 3x192x192 |  tfe-remote (int64 ops) | 41 s | 40.6 s | 42 s |
| 3x192x192 | tf | 0.018 s | 0.007 s | 0.006 s |

Obesrvations:
* Converting the frozen graph to tfe graph didn't affect performance
* Int64 operations improve performance mainly for the first (cold-start) prediction.
* The distributed setup is more than 2x slower than the local
* Performance is still around 3 orders of magnitude worse than vanilla TF

**Comment by gdamaskinos** [comment]
Hi, @jvmancuso -- True that; with a quick check this makes the latency for tfe-local-int64 (with large input) drop to 4.8 secs for the 2nd and 3rd prediction. Nevertheless, the large performance gap with vanilla TF remains.

**Comment by yanndupis** [comment]
Hello @gdamaskinos ,

Thank you very much for your great analysis!

We have benchmarked your model with our internal tools on Google Cloud Platform. On each instance we have 36 CPUs. With int64, we are getting lower runtime than in the table above and they are more aligned with the local performance your are getting.

- **Small input (3x32x32) - tfe-remote (int64 ops)**

<img width="459" alt="screen shot 2018-12-20 at 10 54 41 am" src="https://user-images.githubusercontent.com/8001016/50277855-d556c980-0445-11e9-8f53-80cc1f713a71.png">

- **Large input (3x192x192) - tfe-remote (int64 ops)**
<img width="466" alt="screen shot 2018-12-20 at 10 54 53 am" src="https://user-images.githubusercontent.com/8001016/50277950-1c44bf00-0446-11e9-8ee0-58e284775569.png">

We are observing these differences potentially because we are not using the same machines/ number of CPUs or you have more latency in the network.

In terms of next steps we are planning to:
- Release docs and tools after the holiday to help external contributors benchmarking models with tf-encrypted.
- We believe we could get these numbers down. By inspecting the TensorFlow graph for your model, we have identified several potential optimizations. Several operations could be better parallelized, better memory usage etc. We have listed the issues here #364.
- As mentioned in my earlier message for a variant of vgg16, we were getting a runtime of `1 min 4 seconds` with the securenn protocol. We also benchmarked this model with the POND protocol where we used AveragePooling instead of MaxPooling and an approximated ReLU. The runtime was `4 seconds`. However, the accuracy wasn't good because we accumulated too much error in very large models with the approximated ReLU. We are currently investigating different approximation strategies for activation functions to get good accuracy and fast runtime with the POND protocol. It would be another alternative where we could get very good performance in privacy-preserving setting by slightly adapting the models.

**Comment by avitalsh** [comment]
@yanndupis Hi,
When will you release the benchmarking tools and docs? 

Thank you

**Comment by mortendahl** [comment]
@avitalsh is this still something that's interesting for you?

**Comment by avitalsh** [comment]
@mortendahl Yes!
